core/io/linux.rs

```diff
@@ -1,41 +1,72 @@
 use super::{Completion, File, WriteCompletion, IO};
 use anyhow::Result;
+use libc::iovec;
 use log::trace;
-use std::cell::RefCell;
-use std::os::unix::fs::OpenOptionsExt;
+use std::cell::{Ref, RefCell};
+use nix::fcntl::{self, FcntlArg, OFlag};
 use std::os::unix::io::AsRawFd;
 use std::rc::Rc;
 
+const MAX_IOVECS: usize = 128;
+
 pub struct LinuxIO {
-    ring: Rc<RefCell<io_uring::IoUring>>,
+    inner: Rc<RefCell<InnerLinuxIO>>,
+}
+
+pub struct InnerLinuxIO {
+    ring: io_uring::IoUring,
+    iovecs: [iovec; MAX_IOVECS],
+    next_iovec: usize,
 }
 
 impl LinuxIO {
     pub fn new() -> Result<Self> {
-        let ring = io_uring::IoUring::new(128)?;
+        let ring = io_uring::IoUring::new(MAX_IOVECS as u32)?;
+        let inner = InnerLinuxIO {
+            ring: ring,
+            iovecs: [iovec {
+                iov_base: std::ptr::null_mut(),
+                iov_len: 0,
+            }; MAX_IOVECS],
+            next_iovec: 0,
+        };
         Ok(Self {
-            ring: Rc::new(RefCell::new(ring)),
+            inner: Rc::new(RefCell::new(inner)),
         })
     }
 }
 
+impl InnerLinuxIO {
+    pub fn get_iovec<'a>(&'a mut self, buf: *const u8, len: usize) -> &'a iovec {
+        let iovec = &mut self.iovecs[self.next_iovec];
+        iovec.iov_base = buf as *mut std::ffi::c_void;
+        iovec.iov_len = len;
+        self.next_iovec = (self.next_iovec + 1) % MAX_IOVECS;
+        iovec
+    }
+}
+
 impl IO for LinuxIO {
     fn open_file(&self, path: &str) -> Result<Rc<dyn File>> {
         trace!("open_file(path = {})", path);
         let file = std::fs::File::options()
             .read(true)
             .write(true)
-            .custom_flags(libc::O_DIRECT)
             .open(path)?;
+        // Let's attempt to enable direct I/O. Not all filesystems support it
+        // so ignore any errors.
+        let fd = file.as_raw_fd();
+        nix::fcntl::fcntl(fd, FcntlArg::F_SETFL(OFlag::O_DIRECT))?;
         Ok(Rc::new(LinuxFile {
-            ring: self.ring.clone(),
+            io: self.inner.clone(),
             file,
         }))
     }
 
     fn run_once(&self) -> Result<()> {
         trace!("run_once()");
-        let mut ring = self.ring.borrow_mut();
+        let mut inner = self.inner.borrow_mut();
+        let mut ring = &mut inner.ring;
         ring.submit_and_wait(1)?;
         while let Some(cqe) = ring.completion().next() {
             let c = unsafe { Rc::from_raw(cqe.user_data() as *const Completion) };
```

Comment:
we should check the `cqe.result()` here. I can do later if needed

Reply:
@JoanFM Sure, let's improve error handling in a follow up PR.